source/common/upstream/maglev_lb.cc

```diff
@@ -54,26 +139,88 @@ MaglevTable::MaglevTable(const NormalizedHostWeightVector& normalized_host_weigh
       table_index++;
     }
   }
+}
+CompactMaglevTable::CompactMaglevTable(const NormalizedHostWeightVector& normalized_host_weights,
+                                       double max_normalized_weight, uint64_t table_size,
+                                       bool use_hostname_for_hashing,
+                                       MaglevLoadBalancerStats& stats)
+    : MaglevTable(table_size, stats),
+      table_(absl::bit_width(normalized_host_weights.size()), table_size) {
+  constructMaglevTableInternal(normalized_host_weights, max_normalized_weight,
+                               use_hostname_for_hashing);
+}
 
-  uint64_t min_entries_per_host = table_size_;
-  uint64_t max_entries_per_host = 0;
+void CompactMaglevTable::constructImplementationInternals(
+    std::vector<TableBuildEntry>& table_build_entries, double max_normalized_weight) {
+  // Populate the host table. Index into table_build_entries[i] will align with
+  // the host here.
+  host_table_.reserve(table_build_entries.size());
   for (const auto& entry : table_build_entries) {
-    min_entries_per_host = std::min(entry.count_, min_entries_per_host);
-    max_entries_per_host = std::max(entry.count_, max_entries_per_host);
+    host_table_.emplace_back(entry.host_);
   }
-  stats_.min_entries_per_host_.set(min_entries_per_host);
-  stats_.max_entries_per_host_.set(max_entries_per_host);
+  host_table_.shrink_to_fit();
 
-  if (ENVOY_LOG_CHECK_LEVEL(trace)) {
-    for (uint64_t i = 0; i < table_.size(); i++) {
-      const absl::string_view key_to_hash = hashKey(table_[i], use_hostname_for_hashing);
-      ENVOY_LOG(trace, "maglev: i={} address={} host={}", i, table_[i]->address()->asString(),
-                key_to_hash);
+  // Vector to track whether or not a given fixed width bit is set in the
+  // BitArray used as the maglev table.
+  std::vector<bool> occupied(table_size_, false);
+
+  // Iterate through the table build entries as many times as it takes to fill up the table.
+  uint64_t table_index = 0;
```

Comment:
Since most of the code here is repeated for both original and compact maglev tables, what do you think about my previous comment about wrapping just `table_` and `host_table_` in a class for compact representation, and `table_` from the original maglev table for the "original" representation? 

Will need a solution for tracking "occupied" slots though. You could just keep the `occupied` object in the combined implementation of the algorithm.

Reply:
@jmarantz What do you think? (since you've been tagged as a reviewer).

Reply:
Based on my discussion with @KBaichoo offline, I'm okay with having a TODO for now to evaluate this again at a later date.

Reply:
sg, added a TODO